doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,333 @@
+We expect to reduce the main namespace by a large number of entries, on the
+order of 100. Here is a representative set of examples:
+
+- ``np.inf`` and ``np.nan`` have 8 aliases between them, of which most can be removed.
+- A collection of random and undocumented functions (e.g., ``byte_bounds``, ``disp``,
+  ``safe_eval``, ``who``) listed in
+  `gh-12385 <https://github.com/numpy/numpy/issues/12385>`__
+  can be deprecated and removed.
+- All ``*sctype`` functions can be deprecated and removed, they (see
+  `gh-17325 <https://github.com/numpy/numpy/issues/17325>`__,
+  `gh-12334 <https://github.com/numpy/numpy/issues/12334>`__,
+  and other issues for ``maximum_sctype`` and related functions).
+- Business day functionality can likely be removed (unclear if it needs
+  splitting out like was done for ``np.financial``).
+- The ``np.compat`` namespace, used during the Python 2 to 3 transition, will be removed.
+- Functions that are narrow in scope, with very few public use-cases,
+  will be removed.  See, e.g.
+  ``real_if_close`` (`gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
+  These will have to be identified manually and by issue triage.
+
+New namespaces are introduced for warnings/exceptions (``np.exceptions``) and
+for dtype-related functionality (``np.types``). NumPy 2.0 is a good opportunity
+to populate these submodules from the main namespace.
+
+Functionality that is widely used but has a preferred alternative may either be
+deprecated (with the deprecation message pointing out what to use instead) or
+be hidden by not including it in ``__dir__``. In case of hiding, a ``..
+legacy::`` directory may be used to mark such functionality in the
+documentation.
+
+A test will be added to ensure limited future growth of all namespaces; i.e.,
+every new entry will need to be explicitly added to an allow-list.
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
```

Comment:
A number of these namespaces still have a public submodules such as the likes of `fft.helper`, `linalg.linalg`, `random.bit_generator`, _etc._. Considering they're not explicitly mentioned in any of the sections here, what's the planned future for them? Keep them? Remove them? Deprecate them?

Reply:
I don't consider any of those namespaces public. They're only public if they're explicitly used in the API reference docs. For everything else, https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py#L114-L125 and the lists below that capture their status pretty well.

So yes, remove them all by prepending their names with an underscore.